server/src/main/java/org/elasticsearch/common/util/BitArray.java

```diff
@@ -37,16 +37,32 @@ public BitArray(int initialSize, BigArrays bigArrays) {
         this.bits = bigArrays.newLongArray(initialSize, true);
     }
 
+    /**
+     * Set the {@code index}th bit.
+     */
     public void set(int index) {
         fill(index, true);
     }
 
+    /**
+     * Clear the {@code index}th bit.
+     */
     public void clear(int index) {
         fill(index, false);
     }
 
+    /**
+     * Is the {@code index}th bit set?
+     */
     public boolean get(int index) {
         int wordNum = index >> 6;
+        if (wordNum >= bits.size()) {
+            /*
+             * If the word is bigger than the array then it could *never* have
+             * been set.
+             */
+            return false;
+        }
```

Comment:
That's a good catch. It feels more natural and robust to test BitArray size within the BitArray implementation than add it to the  calling methods (such as in BucketSort). 

When it comes to performance, we can't really say if this test affects some compiler optimization unless we test/benchmark it. Also, I wonder if this is faster or just catching the `ArrayIndexOutOfBoundsException` thrown by `bits.get(wordNum)` within the method and simply returning false. I know try/catch blocks can have performance impacts too.

Reply:
I can check that, but I expect it'll depend on the frequency of `throw`. I already have some benchmarks i can run it against though so it is pretty quick to test.

Reply:
I would assume that most of the times an exception is not thrown. Yet, a try-catch block has its performance implications even no exceptions are thrown.